src/kvstore.h

```diff
@@ -0,0 +1,81 @@
+#ifndef DICTARRAY_H_
+#define DICTARRAY_H_
+
+#include "dict.h"
+#include "adlist.h"
+
+typedef struct {
+    list *rehashing;                       /* List of dictionaries in this kvstore that are currently rehashing. */
+    int resize_cursor;                     /* Cron job uses this cursor to gradually resize dictionaries. */
+    int non_empty_dicts;                   /* The number of non-empty dicts. */
+    unsigned long long key_count;          /* Total number of keys in this kvstore. */
+    unsigned long long bucket_count;       /* Total number of buckets in this kvstore across dictionaries. */
+    unsigned long long *dict_size_index;   /* Binary indexed tree (BIT) that describes cumulative key frequencies up until given dict-index. */
+} kvstoreState;
+
+typedef struct {
+    dictType dtype;
+    dict **dicts;
+    long long num_dicts;
+    long long num_dicts_bits;
+    kvstoreState state;
+} kvstore;
+
+/* Dict metadata for database, used for record the position in rehashing list. */
+typedef struct {
+    kvstore *kvs;
```

Comment:
correct

Reply:
for our knowledge, can you calculate the diff when taking internal frag into account?
@madolson WDYT?

Reply:
I don't really want to pay an extra 200kb just because? What is the material value of adding this as opposed to the previous implementation.

Reply:
in the previous implementation Redis itself managed the `rehashing` list of rehashing dicts, and thus it could trigger the rehashing-related functions in the correct `kvstore` (either main or volatiles)

we want all `kvstore`'s logic to be self-contained, so we need each dict to have a reference to its `kvstore` instance

having said that, i'm open to suggestions, maybe there's a way to keep everything contained to kvstore.c without the 200k additional overhead

Reply:
maybe the rehashing list should be global?
i.e. declared in the kvstore struct as static?
that would solve it, right?

Reply:
it's not only the rehshing list, it also needs to update `bucket_count` whenever a dict is done rehashing

Reply:
ohh, before that we had different callback for each, and the callback knew which counter to increment.
well, maybe we can do that here too somehow.
(not suggesting to generate code on the heap this time :stuck_out_tongue: )
the easiest way would be to add a callback user_data, and add it to the dictType struct.
this way, each kvstore instance, will have a different cloned dictType struct (actually it already does), and the user_data pointer in that struct will be a pointer to the kvstore instance.
so instead of 8 extra bytes per dict, we have 8 extra bytes per kvstore.